commons/all/src/main/java/org/infinispan/commons/netty/BranchlessParser.java

```diff
@@ -0,0 +1,196 @@
+package org.infinispan.commons.netty;
+
+import io.netty.buffer.ByteBuf;
+
+final class BranchlessParser {
+
+   private BranchlessParser() { }
+
+   private static int readableBytes(ByteBuf buffer) {
+      return buffer.writerIndex() - buffer.readerIndex();
+   }
+
+   /**
+    * Reads the buffer for an integer value encoded as Varint.
+    *
+    * <p>
+    * This method tries to read the data from the underlying Netty buffer. If there is not enough bytes to decode the
+    * number, the method returns <code>0</code>. To distinguish between an actual value, the caller must verify the
+    * buffer's reader index. If the index does not change, not data was consumed.
+    * </p>
+    *
+    * <b>Warning:</b> Check the reader index to verify if data was consumed.
+    *
+    * @param buffer The buffer containing the data.
+    * @return The encoded int, or <code>0</code>, if not enough bytes are available.
+    */
+   public static int readRawVarint32(ByteBuf buffer) {
+      if (readableBytes(buffer) < Integer.BYTES)
```

Comment:
This is the only type of check for the index that we should need. Everywhere else in the code should just try to read the bytes and let the exception propagate instead.